### PR TITLE
feat(YouTube - Description components): Add  `Hide 'Key concepts' section` option

### PR DESCRIPTION
**Pull request by ILoveOpenSourceApplications**
https://github.com/ReVanced/revanced-patches/pull/3486

### CODE REVIEW

File: app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java [hunk 1]
```diff
@@ -45,20 +44,42 @@ public DescriptionComponentsFilter() {
                 "transcript_section"
         );
 
+        macroMarkersCarousel = new StringFilterGroup(
+                null,
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:48]: Why is this null and the two groups below not?

Comment (ILoveOpenSourceApplications) [reply app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:48]: `macro_markers_carousel.eml` is the common string which hides both chapters and key concepts having bytearray `chapters_horizontal_shelf` and `learning_concept_macro_markers_carousel_shelf`. So if I use a setting value for macroMarkerCarousel, wouldn't it restrict to either one of the bytearray?

I'm not sure myself, feel free to suggest any required changes as you feel.

File: app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java [hunk 2]
```diff
@@ -45,20 +44,42 @@ public DescriptionComponentsFilter() {
...
+                transcriptSection,
+                macroMarkersCarousel
         );
     }
 
-
     @Override
     boolean isFiltered(@Nullable String identifier, String path, byte[] protobufBufferArray,
                        StringFilterGroup matchedGroup, FilterContentType contentType, int contentIndex) {
-        if (exceptions.matches(path)) return false;
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:61]: Why were the exceptions removed?

Comment (ILoveOpenSourceApplications) [reply app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:61]: Removed by accident. Added it back in b9656614d1b51e20b9a30928e139577b11f9766d

File: app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java [hunk 3]
```diff
@@ -45,21 +44,44 @@ public DescriptionComponentsFilter() {
...
 
-
     @Override
     boolean isFiltered(@Nullable String identifier, String path, byte[] protobufBufferArray,
                        StringFilterGroup matchedGroup, FilterContentType contentType, int contentIndex) {
         if (exceptions.matches(path)) return false;
+        if (matchedGroup == macroMarkersCarousel) {
+            if (contentIndex == 0
+                    && chaptersSectionBufferGroup.check(protobufBufferArray).isFiltered()
+                    && keyConceptsSectionBufferGroup.check(protobufBufferArray).isFiltered()) {
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:78]: Do both options have to be enabled?

File: app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java [hunk 4]
```diff
@@ -45,21 +44,44 @@ public DescriptionComponentsFilter() {
...
+                macroMarkersCarousel
         );
     }
 
-
     @Override
     boolean isFiltered(@Nullable String identifier, String path, byte[] protobufBufferArray,
                        StringFilterGroup matchedGroup, FilterContentType contentType, int contentIndex) {
         if (exceptions.matches(path)) return false;
+        if (matchedGroup == macroMarkersCarousel) {
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:75]:
```suggestion
        if (exceptions.matches(path)) return false;

        if (matchedGroup == macroMarkersCarousel) {
```

File: app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java (hunk 3 above)
Comment (ILoveOpenSourceApplications) [reply app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:79]: Not sure, does only having one enabled be enough to hide both?

Comment (oSumAtrIX) [reply app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:79]: I don't know what you mean by both, but I do not see why `Settings.HIDE_CHAPTERS_SECTION` should control hiding "Key concepts"

Comment (ILoveOpenSourceApplications) [reply app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:79]: For hiding key concepts, there is `Settings.HIDE_KEY_CONCEPS_SECTION` as well. Both buffergroup needs to be defined right?

Comment (ILoveOpenSourceApplications) [reply app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:79]: It's present in the issue [#3324](https://github.com/ReVanced/revanced-patches/issues/3324)

File: app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java [hunk 5]
```diff
@@ -45,21 +46,44 @@ public DescriptionComponentsFilter() {
...
     @Override
     boolean isFiltered(@Nullable String identifier, String path, byte[] protobufBufferArray,
                        StringFilterGroup matchedGroup, FilterContentType contentType, int contentIndex) {
         if (exceptions.matches(path)) return false;
 
+        if (matchedGroup == macroMarkersCarousel) {
+            if (contentIndex == 0)
+                return false;
+        }
+                  if (!macroMarkersCarouselGroupList.check(protobufBufferArray).isFiltered()) {
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:83]:
```suggestion
            if (!macroMarkersCarouselGroupList.check(protobufBufferArray).isFiltered()) {
```

Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:83]: Why is this check outside of the `macroMarkersCarousel` check?

File: app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java [hunk 6]
```diff
@@ -45,21 +46,44 @@ public DescriptionComponentsFilter() {
...
 
-
     @Override
     boolean isFiltered(@Nullable String identifier, String path, byte[] protobufBufferArray,
                        StringFilterGroup matchedGroup, FilterContentType contentType, int contentIndex) {
         if (exceptions.matches(path)) return false;
 
+        if (matchedGroup == macroMarkersCarousel) {
+            if (contentIndex == 0)
+                return false;
```
[the remark this answers is not in the document]
Comment (ILoveOpenSourceApplications) [reply app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java:81]: Applied the requested change.

File: app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java (hunk 5 above)
Comment (ILoveOpenSourceApplications) [reply app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java]: Is it fine now?

Comment (ILoveOpenSourceApplications) [reply app/src/main/java/app/revanced/integrations/youtube/patches/components/DescriptionComponentsFilter.java]: Done.

**Comment by ILoveOpenSourceApplications** [comment]
Also @LisoUseInAIKyrios, adding these below lines require any other changes somewhere?
```kotlin
import app.revanced.integrations.shared.patches.components.ByteArrayFilterGroup;
import app.revanced.integrations.shared.patches.components.ByteArrayFilterGroupList;
```

File: app/src/main/java/app/revanced/integrations/youtube/settings/Settings.java [hunk 7]
```diff
@@ -157,6 +157,7 @@ public class Settings extends BaseSettings {
     public static final BooleanSetting HIDE_ATTRIBUTES_SECTION = new BooleanSetting("revanced_hide_attributes_section", FALSE);
     public static final BooleanSetting HIDE_CHAPTERS_SECTION = new BooleanSetting("revanced_hide_chapters_section", TRUE);
     public static final BooleanSetting HIDE_INFO_CARDS_SECTION = new BooleanSetting("revanced_hide_info_cards_section", TRUE);
+    public static final BooleanSetting HIDE_KEY_CONCEPTS_SECTION = new BooleanSetting("revanced_hide_key_concepts_section", TRUE);
```
[the remark this answers is not in the document]
Comment (ILoveOpenSourceApplications) [reply app/src/main/java/app/revanced/integrations/youtube/settings/Settings.java]: Done.

**Review by oSumAtrIX — APPROVED** [review]

**Pull request merged**